### ISSUE TITLE
Called on non-Map/Set?

### ISSUE THREAD

**Issue by ljharb** [DavidBruant/Map-Set.prototype.toJSON#2]
What happens if this function is borrowed and called on something that's not a Map or a Set?

Should `Map#toJSON` be a generic method that works on anything that provides an `entries` method? Should `Set#toJSON` be a generic method that works on anything iterable?

If not, should they throw, as https://people.mozilla.org/~jorendorff/es6-draft.html#sec-get-map.prototype.size step 3 and https://people.mozilla.org/~jorendorff/es6-draft.html#sec-get-set.prototype.size do?

**Comment by DavidBruant** [comment]
> What happens if this function is borrowed and called on something that's not a Map or a Set?

My initial reaction is to not care too much. These methods are just defaults. They can be overridden if people are unsatisfied with the provided serialization.

Personally, I like the generality of the methods as they are now, but this is obviously open to discussion at this stage.
Do you have reasons or use cases in mind that would justify not having the methods generic as they are now?

**Comment by ljharb** [comment]
I mean, `Map.prototype.toJSON.call(somethingElse)` etc. Typically, these methods are defined in the spec as "intentionally generic", or, as throwing when an internal slot is not found.

I'm not advocating for one or the other, but I would expect it to be part of the specification. Leaving things up to implementations tends to be a future nightmare for the committee :-)

**Comment by DavidBruant** [comment]
Sure. This is still a stage 0 proposal, though :-)
If/when it gets to stage 1, I'll start the spec prose.

**Comment by ljharb** [comment]
I'm writing a polyfill for it now for the `es7-shim` :-) some guidance would be helpful, even without the spec prose.

**Comment by DavidBruant** [comment]
Ahah ok, I wasn't aware of this context :-)
In a way, the code provided doesn't leave too much ambiguity. But it's ES6 code (for-of and ...), so it may not transpile that well (I haven't looked too much at the state of the art of transpiling, but I imagine edge cases may prevent easy transpilation for for-of).

For `Set#toJSON`, I guess the code means what I intended : `this` is the dynamic one and should be iterable and the purpose is just to put in space what iterators iterate over in time. This is a problem if the iterator never ends (if you want to reuse this methods on other objects), but this never happens for Sets by construction. In a way, trying to serialize an infinite iterator is a mistake anyway.

For `Map#toJSON`, I guess `entries` are a thing now in ES6 (used in arrays, typed arrays, sets and maps), so it's fair to make the function generic using it, I think.
I lazily used `Object.create` + `[[Set]]` (in the ES6 `6.1.7.2` sense) to avoid issues with `__proto__` and inherited setters; a more reasonable approach would be to start with `{}` and do `Object.defineProperty`.

Unrelated, but isn't it too early for a polyfill? This is still stage 0, details may change.

**Comment by ljharb** [comment]
My polyfill wouldn't use for..of, it'd use a `while` loop and the iterator `.next()` API, and it would work in ES3 (if Map/Set were polyfillable in ES3)

It makes sense then that `Set#toJSON` just requires that `this[Symbol.iterator]` exists, and then creates an `Array` from it.

I like the empty object and `defineProperty` approach. Instead of using the `entries` method though, perhaps it could use https://people.mozilla.org/~jorendorff/es6-draft.html#sec-map.prototype.entries and `CreateMapIterator(this, "key+value")` internally, and then use that iterator in `for..of`? That would work more generically, and with subclasses, without trying to do the same thing for non-Map things.

Just like the `es6-shim` in its early days, I intend to just keep iterating the `es7-shim` polyfills so they can serve as example implementations :-)

**Comment by DavidBruant** [comment]
> That would work more generically, and with subclasses, without trying to do the same thing for non-Map things.

Subclasses inherit from `Map.prototype` (and `entries` along the way), no?

But I see a point. In ES6, objects may have a `[[MapData]]` without an `.entries()` method. It'd be good for `Map#toJSON` to work on these objects. I'll file a specific issue for this.

> Just like the es6-shim in its early days, I intend to just keep iterating the es7-shim polyfills so they can serve as example implementations :-)

Sounds good.

**Comment by zloirock** [comment]
BTW, about polyfills - I have already added to `core-js` `Set#toJSON`, but not sure about `Map#toJSON` - see #1.

**Comment by zloirock** [comment]
@DavidBruant emm... after https://github.com/DavidBruant/Map-Set.prototype.toJSON/issues/1#issuecomment-94268032 current `Map#toJSON` behavior is:

``` js
Map.prototype.toJSON = function(){
  return [...this];
}
// or:
Map.prototype.toJSON = function(){
  return [...this.entries()]; // or something like it
}
// ?
```

For `Map` `Map.prototype[Symbol.iterator] === Map.prototype.entries`, but not for all other iterables.

Strange in generic iterators-based method use method / behavior missing in iterators protocol.

Using simple `[...this]` as `toJSON` for all iterables allows standardize method `toJSON` for all (most) future iterable collections and add it to `%IteratorPrototype%` object for iterators.

**Comment by ljharb** [comment]
However, it means that `Map#toJSON` wouldn't reliably produce an array of entries suitable for passing into `new Map(entries)` - I think that'd be a useful property.

**Comment by zloirock** [comment]
@ljharb it can reproduce `Map` from json created from `Map`, the rest is not so important :) Broken iterators protocol is more important problem. How do you propose call it at custom iterators? Possible use something like

``` js
Map.prototype.toJSON = function(){
  var entries = [];
  for(var e of this){
    if(e == null || typeof e != 'object')throw TypeError();
    entries.push(e);
  }
  return entries;
}
```

but it can't be called on, for example, `[1,  2, 3]`, like in your shim, only `[1,  2, 3].entries()`. Simply use more  generic `[...this]` without additional checks and special behavior for some data types.

**Comment by ljharb** [comment]
I don't see as much value in a truly generic `toJSON` function as opposed to one that makes the most sense for `Map`s.

To be honest, I think the most straightforward would be if it was not generic, but instead threw when there wasn't a `[[MapData]]` internal slot.

**Comment by DavidBruant** [comment]
closed via #5
